e2e/internal/e2e/profile/profile.go

```diff
@@ -0,0 +1,230 @@
+// it's missing the In and String methods). This exists in order to
+// reduce code duplication.
+type baseProfile struct{}
+
+func (baseProfile) Privileged() bool {
+	return false
+}
+
+func (baseProfile) Requirements(t *testing.T) {
+}
+
+func (baseProfile) Args([]string) []string {
+	return nil
+}
+
+func (baseProfile) User(t *testing.T) *user.User {
+	return getOrigUser(t)
+}
+
+// User implements the Profile interface describing a non-privileged
+// user.
+type User struct {
+	baseProfile
+}
+
+func (p User) In(profiles ...Profile) bool {
+	return in(p, profiles)
+}
+
+func (User) String() string {
+	return "User"
+}
+
+// Root implements the Profile interface describing a privileged user.
+type Root struct {
+	baseProfile
+}
+
+func (Root) Privileged() bool {
+	return true
+}
+
+func (Root) User(t *testing.T) *user.User {
+	u, err := user.GetPwUID(0)
+	if err != nil {
+		t.Fatalf("failed to retrieve root user information: %s", err)
+	}
+	return u
+}
+
+func (p Root) In(profiles ...Profile) bool {
+	return in(p, profiles)
+}
+
+func (Root) String() string {
+	return "Root"
+}
+
+// Fakeroot implements the Profile interface describing a "fake" root
+// user and makes use of singularity --fakeroot flag where appropriate.
+type Fakeroot struct {
+	baseProfile
+}
+
+func (Fakeroot) Requirements(t *testing.T) {
+	require.UserNamespace(t)
+
+	// check that current user has valid mappings in /etc/subuid
+	if _, err := fakeroot.GetIDRange(fakeroot.SubUIDFile, uint32(e2e.OrigUID())); err != nil {
+		t.Fatalf("fakeroot configuration error: %s", err)
+	}
+
+	// check that current user has valid mappings in /etc/subgid
+	if _, err := fakeroot.GetIDRange(fakeroot.SubGIDFile, uint32(e2e.OrigUID())); err != nil {
+		t.Fatalf("fakeroot configuration error: %s", err)
+	}
+}
+
+func (Fakeroot) Args(cmd []string) []string {
+	commands := []string{
+		"shell",
+		"exec",
+		"run",
+		"test",
+		"instance start",
+		"build",
+	}
+
+	return buildArgs(cmd, commands, "--fakeroot")
+}
+
+func (Fakeroot) User(t *testing.T) *user.User {
+	return getOrigUser(t)
+}
+
+func (p Fakeroot) In(profiles ...Profile) bool {
+	return in(p, profiles)
+}
+
+func (Fakeroot) String() string {
+	return "Fakeroot"
+}
+
+// UserNamespace implemets Profile describing an unprivileged user that
+// operates in a separate user namespace.
+type UserNamespace struct {
+	baseProfile
+}
+
+func (UserNamespace) Args(cmd []string) []string {
+	commands := []string{
+		"shell",
+		"exec",
+		"run",
+		"test",
+		"instance start",
+	}
+
+	return buildArgs(cmd, commands, "--userns")
+}
+
+func (UserNamespace) Requirements(t *testing.T) {
+	require.UserNamespace(t)
+}
+
+func (p UserNamespace) In(profiles ...Profile) bool {
+	return in(p, profiles)
+}
+
+func (UserNamespace) String() string {
+	return "UserNamespace"
+}
+
+// RootUserNamespace implemets Profile describing a privileged user that
+// operates in a separate user namespace.
+type RootUserNamespace struct {
```

Comment:
Honestly, I don't really like this double embedding. It took some time to determine why `RootUserNamespace .Privileged()` will refer `Root.Privileged()`. 
This is diamond embedding. I understand, that this may be convenient, but really hard to understand at the first glance.

Reply:
Yeah, I had a different version of this with a single embedded struct. Then I switched to embedding the other. And then I ended up embedding both because that's what RootUserNamespace is: one part UserNamespace, one part Root, with Root defining the user and UserNamespace defining flags and requirements. The "purely imperative" version of this (for lack of a better way to describe it) is worse, because it has different switch cases depending on the function you are looking at.

The subtlety here is that these structs do not have overlapping methods. If both structs where directly defining the same method, say Privileged, you wouldn't be able to call RootUserNamespace.Privileged() without specifying which one you want (or without implementing Privileged directly for RootUserNamespace). That's Go's solution to the diamond problem: do not allow the ambiguity at the call location.

So, while I agree with you, I'm not sure what to do about it...